blockchain/thresholdstate.go

```diff
@@ -0,0 +1,336 @@
+				state = ThresholdFailed
+				break
+			}
+
+			// The state for the rule moves to the started state
+			// once its start time has been reached (and it hasn't
+			// already expired per the above).
+			if medianTimeUnix >= checker.BeginTime() {
+				state = ThresholdStarted
+			}
+
+		case ThresholdStarted:
+			// The deployment of the rule change fails if it expires
+			// before it is accepted and locked in.
+			medianTime, err := b.calcPastMedianTime(prevNode)
+			if err != nil {
+				return ThresholdFailed, err
+			}
+			if uint64(medianTime.Unix()) >= checker.EndTime() {
+				state = ThresholdFailed
+				break
+			}
+
+			// At this point, the rule change is still being voted
+			// on by the miners, so iterate backwards through the
+			// confirmation window to count all of the votes in it.
+			var count uint32
+			countNode := prevNode
+			for i := int32(0); i < confirmationWindow; i++ {
+				condition, err := checker.Condition(countNode)
+				if err != nil {
+					return ThresholdFailed, err
+				}
+				if condition {
+					count++
+				}
+
+				// Get the previous block node.  This function
+				// is used over simply accessing countNode.parent
+				// directly as it will dynamically create
+				// previous block nodes as needed.  This helps
+				// allow only the pieces of the chain that are
+				// needed to remain in memory.
+				countNode, err = b.getPrevNodeFromNode(countNode)
+				if err != nil {
+					return ThresholdFailed, err
+				}
+			}
+
+			// The state is locked in if the number of blocks in the
+			// period that voted for the rule change meets the
+			// activation threshold.
+			if count >= checker.RuleChangeActivationThreshold() {
+				state = ThresholdLockedIn
+			}
+
+		case ThresholdLockedIn:
+			// The new rule becomes active when its previous state
+			// was locked in.
+			state = ThresholdActive
+
+		// Nothing to do if the previous state is active or failed since
+		// they are both terminal states.
+		case ThresholdActive:
+		case ThresholdFailed:
+		}
+
+		// Update the cache to avoid recalculating the state in the
+		// future.
+		cache.Update(*prevNode.hash, state)
+	}
+
+	return state, nil
+}
+
+// ThresholdState returns the current rule change threshold state of the given
+// deployment ID for the end of the current best chain.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) ThresholdState(deploymentID uint32) (ThresholdState, error) {
+	b.chainLock.Lock()
+	state, err := b.thresholdForDeployment(b.bestNode, deploymentID)
+	b.chainLock.Unlock()
+
+	return state, err
+}
+
+// thresholdForDeployment returns the current rule change threshold for a
+// given deploymentID. The threshold is evaluated from the point of view of the
+// blockNode passed in as the first argument to this method.
+//
+// This function MUST be called with the chain state lock held (for writes).
+func (b *BlockChain) thresholdForDeployment(blockNode *blockNode,
```

Comment:
I don't want to bikeshed too much since it's important we get this PR finished and landed, but I wonder if this might be better named `deploymentState` since I noticed that `thresholdForDeployment` is quite long and most of the callers end up using same variant of the name `state` in the variable anyways.  Definitely not something that has to be changed, but if you're in there making modifications anyways to fix the actual code issues pointed out regarding reorgs, and you think it's a good idea, go for it.